### PR TITLE
Add helpers for txpool

### LINKED ISSUE #72: Add helpers for fee currency handling
In the txpool we need a new interface that provides information required for handling fee currencies.

`FeeContext`:
- `IsWhitelisted(feeCurrency address): bool`
- `GetBalanceOf(account address, feeCurrency address): int`
- `CompareFees(fee1 int, feCurrency1 address, fee2 int, feeCurrency2 address): int`

@hbandura Please have a look and add things that would help you. Also please improve naming if you have better ideas :)

### LINKED ISSUE #72: L1 data fee and fee currencies
Part of celo-org/optimism#6 

Optimism has the concept of the [l1 data fee](https://community.optimism.io/docs/developers/build/transaction-fees/#the-l1-data-fee). This part of the transaction fee pays for the fees related to L1 submission of L2 transactions, i.e. L1 data available. It is calculated as follows, where`l1_gas_price` is an external input which needs to be made available.
```
l1_data_fee = l1_gas_price * (tx_data_gas + fixed_overhead) * dynamic_overhead
```

The total transaction fee can be calculated by adding the l1 data fee to the l2 execution fee.

## Fee currencies

As the l1 data fee is denominated in ETH, there needs to be a way to convert it into the fee currency.
If the fee currency is:
- CELO: Use the exchange rate available from celo-org/optimism#55 and use it to convert the l1 data fee from ETH to CELO
- ERC20: Use the exchange rate available from celo-org/optimism#55 and use it to convert the l1 data fee from ETH to CELO, and then the fee currency exchange rate available from oracles to convert it into the respective ERC20 token.

### PR DESCRIPTION
**Pull request by palango**
Resolves [#72](https://github.com/celo-org/optimism/issues/72)

Adds a new `CeloContext` as described in [#72](https://github.com/celo-org/optimism/issues/72). Will need changes after #26 .

### Open questions
- ~~What should we do if a fee currency is whitelisted but we cannot get an exchange rate?~~
- Can we use the exchange rates available in the `BlockContext`?

### CODE REVIEW

**Comment by karlb** [comment]
> What should we do if a fee currency is whitelisted but we cannot get an exchange rate?

Is there any real choice but not to include txs with that fee currency?

**Comment by palango** [comment]
> > What should we do if a fee currency is whitelisted but we cannot get an exchange rate?
> 
> Is there any real choice but not to include txs with that fee currency?

I'm asking because this is not what we currently do: https://github.com/celo-org/celo-blockchain/blob/d2e7c9f3a44a727a8117da9c928a3ccb8e582f27/contracts/currency/currency.go#L206

But this behaviour doesn't make sense to me either and I think we should avoid those transactions.

**Comment by hbandura** [comment]
Maybe the "IsWhitelisted" should be something like "IsAvailable", and if we cannot find the exchange rate OR if it's not whitelisted, we should drop all txs

**Comment by palango** [comment]
> Maybe the "IsWhitelisted" should be something like "IsAvailable", and if we cannot find the exchange rate OR if it's not whitelisted, we should drop all txs

It behaves like that now. But I still think `IsWhitelisted` fit's better.

File: core/txpool/legacypool/celo.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+)
+
+var (
+	unitRate = big.NewRat(1, 1)
+)
+
+type celoContext interface {
+	IsWhitelisted(feeCurrency common.Address) bool
+	GetBalanceOf(account common.Address, feeCurrency common.Address) (*big.Int, error)
+	CompareValue(fee1 *big.Int, feCurrency1 common.Address, fee2 *big.Int, feeCurrency2 common.Address) int
```
Comment (carterqw2) [inline core/txpool/legacypool/celo.go]:
```suggestion
	CompareValue(fee1 *big.Int, feeCurrency1 common.Address, fee2 *big.Int, feeCurrency2 common.Address) int
```

File: core/celo_evm.go [hunk 2]
```diff
@@ -16,7 +16,7 @@ import (
 )
 
 // Returns the exchange rates for all gas currencies from CELO
-func getExchangeRates(caller *CeloBackend) (map[common.Address]*big.Rat, error) {
+func GetExchangeRates(caller *CeloBackend) (common.ExchangeRates, error) {
```
Comment (carterqw2) [inline core/celo_evm.go]: Why not change the type to `ContractCaller` interface here?

File: core/txpool/legacypool/celo_test.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			cc := &celoContextImpl{
+				backend:       nil,
+				exchangeRates: exchangeRates,
+			}
+			got, err := cc.CompareValue(tt.args.val1, tt.args.feeCurrency1, tt.args.val2, tt.args.feeCurrency2)
+
+			if tt.wantErr && err == nil {
+				t.Error("Expected error in celoCOntextImpl.CompareFees")
```
Comment (carterqw2) [inline core/txpool/legacypool/celo_test.go]:
```suggestion
				t.Error("Expected error in celoContextImpl.CompareFees")
```

File: core/txpool/legacypool/celo.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+type celoContextImpl struct {
+	backend       *core.CeloBackend
+	exchangeRates common.ExchangeRates
+}
+
+func newCeloContext(config *params.ChainConfig, state *state.StateDB) *celoContextImpl {
+	backend := &core.CeloBackend{
```
Comment (carterqw2) [inline core/txpool/legacypool/celo.go]: Perhaps we should inject it as a dependency instead?

Comment (palango) [reply core/txpool/legacypool/celo.go]: Good point, change it.

**Comment by palango** [comment]
Included in #43

**Pull request closed without merging**